src/debug.h

```diff
@@ -1,14 +1,167 @@
+    // outside the debug header.
+    // NOTE: intentionally non-consecutive values.
+    typedef enum { 
+        E_DEBUG_LEVEL_FATAL   = 0x00u,
+        E_DEBUG_LEVEL_ERROR   = 0x20u,
+        E_DEBUG_LEVEL_WARNING = 0x60u,
+        E_DEBUG_LEVEL_INFO    = 0xA0u,
+        E_DEBUG_LEVEL_VERBOSE = 0xC0u,
+        E_DEBUG_LEVEL_DEBUG   = 0xE0u,
+        E_DEBUG_LEVEL_NEVER   = 0xFFu,
+    } _bp_debug_level_enum_t;
+    typedef enum _bp_debug_category_enum_t {
+        E_DEBUG_CAT_CATCHALL         =  0u, // (((uint32_t)1u) <<  0u), // for messages that are not (yet) categorized
+        E_DEBUG_CAT_EARLY            =  1u, // (((uint32_t)1u) <<  1u), // early-in-boot (initialization)
+        E_DEBUG_CAT_ONBOARD_PIXELS   =  2u, // (((uint32_t)1u) <<  2u), // onboard RGB pixels
+        E_DEBUG_CAT_USB_HID          =  3u, // (((uint32_t)1u) <<  3u), // USB based HID interactions
+        E_DEBUG_CAT_USB_CDC          =  4u, // (((uint32_t)1u) <<  4u), // USB based serial port
+        E_DEBUG_CAT_USB_MSC          =  5u, // (((uint32_t)1u) <<  5u), // USB based mass storage commands
+        E_DEBUG_CAT_FLASH_MEDIA      =  6u, // (((uint32_t)1u) <<  6u), // flash media operations
+        E_DEBUG_CAT_UI               =  7u, // (((uint32_t)1u) <<  7u), // for files in `src/ui/...`
+        E_DEBUG_CAT_BINMODE_FALA     =  8u, // (((uint32_t)1u) <<  8u), // follow-along logic analyzer
+        E_DEBUG_CAT_BINMODE_SUMP     =  9u, // (((uint32_t)1u) <<  9u), // SUMP mode (logic analyzer)
+        E_DEBUG_CAT_BYTECODE         = 10u, // (((uint32_t)1u) << 10u), // aka syntax parsing into bytecode, bytecode execution
+        E_DEBUG_CAT_MODE_GLOBAL      = 11u, // (((uint32_t)1u) << 11u), // global mode parsing
+        E_DEBUG_CAT_MODE_1WIRE       = 12u, // (((uint32_t)1u) << 12u), // 1-wire mode specific
+        E_DEBUG_CAT_MODE_2WIRE       = 13u, // (((uint32_t)1u) << 13u), // 2-wire mode specific
+        E_DEBUG_CAT_MODE_UART        = 14u, // (((uint32_t)1u) << 14u), // uart mode specific
+        E_DEBUG_CAT_MODE_I2C         = 15u, // (((uint32_t)1u) << 15u), // i2c mode specific
+        E_DEBUG_CAT_MODE_SPI         = 16u, // (((uint32_t)1u) << 16u), // spi mode specific
+        E_DEBUG_CAT_MODE_DIO         = 17u, // (((uint32_t)1u) << 17u), // dio mode specific
+        E_DEBUG_CAT_MODE_HWLED       = 18u, // (((uint32_t)1u) << 18u), // hwled mode specific
+        E_DEBUG_CAT_MODE_INFRARED    = 19u, // (((uint32_t)1u) << 19u), // infrared mode specific
+        E_DEBUG_CAT_MODE_BINLOOPBACK = 20u, // (((uint32_t)1u) << 20u), // binloopback mode specific
+        E_DEBUG_CAT_MODE_LCDSPI      = 21u, // (((uint32_t)1u) << 21u), // lcdspi mode specific
+        E_DEBUG_CAT_MODE_LCDI2C      = 22u, // (((uint32_t)1u) << 22u), // lcdi2c mode specific
+        // ... reserved for future use ...
+        E_DEBUG_CAT_TEMP2            = 31u, // (((uint32_t)1u) << 31u), // use for temporary debug messages
```

Comment:
Thanks, good to note.

I added explicit comments in the code around this.   Without moving to a (embedded focused subset of) C++, there's not much way to statically enforce this at compilation time.  Thus, since it only affects developers, I am hoping the comments to be sufficient.

Note that, for the actual categories ... I intend to comment those out initially.  I added the strawman categories to get a "feel" for how terrible it would be to add a new category.